IDE0270 creates invalid code

**Version Used**: 
Version 17.11.0 Preview 3.0 [35025.322.main]

**Steps to Reproduce**:
```C#
static void Repro(object? o)
{
    IList<int>? list = o as int[];
    if (list is null)
    {
        list = o as List<int>;
    }
}
``` 

**Diagnostic Id**:
IDE0270

**Expected Behavior**:
Diagnostic is not triggered if the resulting fixer's code won't compile.

**Actual Behavior**:
Diagnostic fires, and the fixer generates this code:
```C#
static void Repro(object? o)
{
    IList<int>? list = o as int[] ?? o as List<int>;
}
```
which fails to compile with:
```
error CS0019: Operator '??' cannot be applied to operands of type 'int[]' and 'List<int>'
```

--- a/src/Analyzers/CSharp/CodeFixes/CSharpCodeFixes.projitems
+++ b/src/Analyzers/CSharp/CodeFixes/CSharpCodeFixes.projitems
@@ -77,6 +77,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)UnsealClass\CSharpUnsealClassCodeFixProvider.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UpdateProjectToAllowUnsafe\CSharpUpdateProjectToAllowUnsafeCodeFixProvider.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UpgradeProject\CSharpUpgradeProjectCodeFixProvider.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)UseCoalesceExpression\CSharpUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UseCollectionExpression\CSharpCollectionExpressionRewriter.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UseCollectionExpression\CSharpUseCollectionExpressionForBuilderCodeFixProvider.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UseCollectionExpression\CSharpUseCollectionExpressionForFluentCodeFixProvider.cs" />

--- a/src/Analyzers/CSharp/CodeFixes/UseCoalesceExpression/CSharpUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
+++ b/src/Analyzers/CSharp/CodeFixes/UseCoalesceExpression/CSharpUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
@@ -1,0 +1,57 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageService;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.UseCoalesceExpression;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseCoalesceExpression;
+
+[ExtensionOrder(Before = PredefinedCodeFixProviderNames.AddBraces)]
+[ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.UseCoalesceExpressionForIfNullStatementCheck), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class CSharpUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider()
+    : AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider
+{
+    protected override bool ShouldAddExplicitCast(
+        ISyntaxFactsService syntaxFacts, SemanticModel semanticModel,
+        SyntaxNode expressionToCoalesce, SyntaxNode whenTrueStatement,
+        [NotNullWhen(true)] out ITypeSymbol? castTo,
+        CancellationToken cancellationToken)
+    {
+        castTo = null;
+
+        if (!syntaxFacts.IsSimpleAssignmentStatement(whenTrueStatement))
+            return false;
+
+        syntaxFacts.GetPartsOfAssignmentStatement(whenTrueStatement, out var left, out var right);
+
+        var leftPartTypeSymbol = semanticModel.GetTypeInfo(expressionToCoalesce, cancellationToken).Type;
+        var rightPartTypeSymbol = semanticModel.GetTypeInfo(right, cancellationToken).Type;
+        var finalDestinationTypeSymbol = semanticModel.GetTypeInfo(left, cancellationToken).Type;
+
+        if (leftPartTypeSymbol == null || rightPartTypeSymbol == null || finalDestinationTypeSymbol == null)
+            return false;
+
+        if (leftPartTypeSymbol.Equals(rightPartTypeSymbol))
+            return false;
+
+        if (semanticModel.Compilation.HasImplicitConversion(leftPartTypeSymbol, rightPartTypeSymbol) ||
+            semanticModel.Compilation.HasImplicitConversion(rightPartTypeSymbol, leftPartTypeSymbol))
+        {
+            return false;
+        }
+
+        castTo = finalDestinationTypeSymbol;
+        return true;
+    }
+}

--- a/src/Analyzers/CSharp/CodeFixes/UseCoalesceExpression/CSharpUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
+++ b/src/Analyzers/CSharp/CodeFixes/UseCoalesceExpression/CSharpUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
@@ -20,18 +20,13 @@ internal sealed class CSharpUseCoalesceExpressionForIfNullStatementCheckCodeFixP
     : AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider
 {
     protected override ITypeSymbol? TryGetExplicitCast(
-        ISyntaxFactsService syntaxFacts, SemanticModel semanticModel,
-        SyntaxNode expressionToCoalesce, SyntaxNode whenTrueStatement,
+        SemanticModel semanticModel, SyntaxNode expressionToCoalesce,
+        SyntaxNode leftAssignmentPart, SyntaxNode rightAssignmentPart,
         CancellationToken cancellationToken)
     {
-        if (!syntaxFacts.IsSimpleAssignmentStatement(whenTrueStatement))
-            return null;
-
-        syntaxFacts.GetPartsOfAssignmentStatement(whenTrueStatement, out var left, out var right);
-
         var leftPartTypeSymbol = semanticModel.GetTypeInfo(expressionToCoalesce, cancellationToken).Type;
-        var rightPartTypeSymbol = semanticModel.GetTypeInfo(right, cancellationToken).Type;
-        var finalDestinationTypeSymbol = semanticModel.GetTypeInfo(left, cancellationToken).Type;
+        var rightPartTypeSymbol = semanticModel.GetTypeInfo(rightAssignmentPart, cancellationToken).Type;
+        var finalDestinationTypeSymbol = semanticModel.GetTypeInfo(leftAssignmentPart, cancellationToken).Type;
 
         if (leftPartTypeSymbol == null || rightPartTypeSymbol == null || finalDestinationTypeSymbol == null)
             return null;

--- a/src/Analyzers/CSharp/CodeFixes/UseCoalesceExpression/CSharpUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
+++ b/src/Analyzers/CSharp/CodeFixes/UseCoalesceExpression/CSharpUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
@@ -4,13 +4,10 @@
 
 using System;
 using System.Composition;
-using System.Diagnostics.CodeAnalysis;
 using System.Threading;
 using Microsoft.CodeAnalysis.CodeFixes;
-using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.LanguageService;
-using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.UseCoalesceExpression;
 
 namespace Microsoft.CodeAnalysis.CSharp.UseCoalesceExpression;
@@ -22,16 +19,13 @@ namespace Microsoft.CodeAnalysis.CSharp.UseCoalesceExpression;
 internal sealed class CSharpUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider()
     : AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider
 {
-    protected override bool ShouldAddExplicitCast(
+    protected override ITypeSymbol? TryGetExplicitCast(
         ISyntaxFactsService syntaxFacts, SemanticModel semanticModel,
         SyntaxNode expressionToCoalesce, SyntaxNode whenTrueStatement,
-        [NotNullWhen(true)] out ITypeSymbol? castTo,
         CancellationToken cancellationToken)
     {
-        castTo = null;
-
         if (!syntaxFacts.IsSimpleAssignmentStatement(whenTrueStatement))
-            return false;
+            return null;
 
         syntaxFacts.GetPartsOfAssignmentStatement(whenTrueStatement, out var left, out var right);
 
@@ -40,18 +34,17 @@ internal sealed class CSharpUseCoalesceExpressionForIfNullStatementCheckCodeFixP
         var finalDestinationTypeSymbol = semanticModel.GetTypeInfo(left, cancellationToken).Type;
 
         if (leftPartTypeSymbol == null || rightPartTypeSymbol == null || finalDestinationTypeSymbol == null)
-            return false;
+            return null;
 
         if (leftPartTypeSymbol.Equals(rightPartTypeSymbol))
-            return false;
+            return null;
 
         if (semanticModel.Compilation.HasImplicitConversion(leftPartTypeSymbol, rightPartTypeSymbol) ||
             semanticModel.Compilation.HasImplicitConversion(rightPartTypeSymbol, leftPartTypeSymbol))
         {
-            return false;
+            return null;
         }
 
-        castTo = finalDestinationTypeSymbol;
-        return true;
+        return finalDestinationTypeSymbol;
     }
 }

--- a/src/Analyzers/CSharp/Tests/UseCoalesceExpression/UseCoalesceExpressionForIfNullStatementCheckTests.cs
+++ b/src/Analyzers/CSharp/Tests/UseCoalesceExpression/UseCoalesceExpressionForIfNullStatementCheckTests.cs
@@ -4,16 +4,17 @@
 
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.CSharp.Analyzers.UseCoalesceExpression;
+using Microsoft.CodeAnalysis.CSharp.UseCoalesceExpression;
 using Microsoft.CodeAnalysis.Editor.UnitTests.CodeActions;
 using Microsoft.CodeAnalysis.Test.Utilities;
-using Microsoft.CodeAnalysis.UseCoalesceExpression;
+using Roslyn.Test.Utilities;
 using Xunit;
 
 namespace Microsoft.CodeAnalysis.CSharp.Analyzers.UnitTests.UseCoalesceExpression;
 
 using VerifyCS = CSharpCodeFixVerifier<
     CSharpUseCoalesceExpressionForIfNullStatementCheckDiagnosticAnalyzer,
-    UseCoalesceExpressionForIfNullStatementCheckCodeFixProvider>;
+    CSharpUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider>;
 
 [Trait(Traits.Feature, Traits.Features.CodeActionsUseCoalesceExpression)]
 public class UseCoalesceExpressionForIfNullStatementCheckTests
@@ -456,6 +457,238 @@ public class UseCoalesceExpressionForIfNullStatementCheckTests
             TestCode = text,
             FixedCode = text,
             LanguageVersion = LanguageVersion.CSharp9,
+        }.RunAsync();
+    }
+
+    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/74460")]
+    public async Task TestLocalDeclaration_CastWithParenthesizedExpression()
+    {
+        await new VerifyCS.Test
+        {
+            TestCode =
+            """
+            interface I
+            {
+            }
+
+            class C : I
+            {
+                void M(object o)
+                {
+                    I item = o as C;
+                    [|if|] (item == null)
+                    {
+                        item = o as D;
+                    }
+                }
+            }
+
+            class D : I
+            {
+            }
+            """,
+            FixedCode =
+            """
+            interface I
+            {
+            }
+
+            class C : I
+            {
+                void M(object o)
+                {
+                    I item = (I)(o as C) ?? o as D;
+                }
+            }
+
+            class D : I
+            {
+            }
+            """
+        }.RunAsync();
+    }
+
+    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/74460")]
+    public async Task TestLocalDeclaration_CastWithoutParenthesizedExpression()
+    {
+        await new VerifyCS.Test
+        {
+            TestCode =
+            """
+            interface I
+            {
+            }
+
+            class C : I
+            {
+                void M(C c, D d)
+                {
+                    I item = c;
+                    [|if|] (item == null)
+                    {
+                        item = d;
+                    }
+                }
+            }
+
+            class D : I
+            {
+            }
+            """,
+            FixedCode =
+            """
+            interface I
+            {
+            }
+
+            class C : I
+            {
+                void M(C c, D d)
+                {
+                    I item = (I)c ?? d;
+                }
+            }
+
+            class D : I
+            {
+            }
+            """
+        }.RunAsync();
+    }
+
+    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/74460")]
+    public async Task TestLocalDeclaration_NoCastWhenEqualSymbol()
+    {
+        await new VerifyCS.Test
+        {
+            TestCode =
+            """
+            interface I
+            {
+            }
+
+            class C : I
+            {
+                void M(C c1, C c2)
+                {
+                    I item = c1;
+                    [|if|] (item == null)
+                    {
+                        item = c2;
+                    }
+                }
+            }
+            """,
+            FixedCode =
+            """
+            interface I
+            {
+            }
+
+            class C : I
+            {
+                void M(C c1, C c2)
+                {
+                    I item = c1 ?? c2;
+                }
+            }
+            """
+        }.RunAsync();
+    }
+
+    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/74460")]
+    public async Task TestLocalDeclaration_NoCastWhenDerivedClass()
+    {
+        await new VerifyCS.Test
+        {
+            TestCode =
+            """
+            interface I
+            {
+            }
+
+            class C : I
+            {
+                void M(C c, D d)
+                {
+                    I item = c;
+                    [|if|] (item == null)
+                    {
+                        item = d;
+                    }
+                }
+            }
+
+            class D : C
+            {
+            }
+            """,
+            FixedCode =
+            """
+            interface I
+            {
+            }
+
+            class C : I
+            {
+                void M(C c, D d)
+                {
+                    I item = c ?? d;
+                }
+            }
+
+            class D : C
+            {
+            }
+            """
+        }.RunAsync();
+    }
+
+    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/74460")]
+    public async Task TestLocalDeclaration_NoCastWhenDerivedClassReversed()
+    {
+        await new VerifyCS.Test
+        {
+            TestCode =
+            """
+            interface I
+            {
+            }
+
+            class C : D
+            {
+                void M(C c, D d)
+                {
+                    I item = c;
+                    [|if|] (item == null)
+                    {
+                        item = d;
+                    }
+                }
+            }
+
+            class D : I
+            {
+            }
+            """,
+            FixedCode =
+            """
+            interface I
+            {
+            }
+
+            class C : D
+            {
+                void M(C c, D d)
+                {
+                    I item = c ?? d;
+                }
+            }
+
+            class D : I
+            {
+            }
+            """
         }.RunAsync();
     }
 }

--- a/src/Analyzers/Core/CodeFixes/CodeFixes.projitems
+++ b/src/Analyzers/Core/CodeFixes/CodeFixes.projitems
@@ -70,7 +70,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)SimplifyInterpolation\AbstractSimplifyInterpolationCodeFixProvider.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)SimplifyLinqExpression\AbstractSimplifyLinqExpressionCodeFixProvider`3.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UpgradeProject\AbstractUpgradeProjectCodeFixProvider.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)UseCoalesceExpression\UseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)UseCoalesceExpression\AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UseCoalesceExpression\UseCoalesceExpressionForTernaryConditionalCheckCodeFixProvider.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UseCoalesceExpression\UseCoalesceExpressionForNullableTernaryConditionalCheckCodeFixProvider.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UseCollectionExpression\AbstractUseCollectionExpressionCodeFixProvider.cs" />

--- a/src/Analyzers/Core/CodeFixes/UseCoalesceExpression/AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
+++ b/src/Analyzers/Core/CodeFixes/UseCoalesceExpression/AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
@@ -58,7 +58,12 @@ internal abstract class AbstractUseCoalesceExpressionForIfNullStatementCheckCode
         SyntaxNode TryAddExplicitCast(SyntaxNode expressionToCoalesce, SyntaxNode whenTrueStatement)
         {
             // This can be either SimpleAssignmentStatement or ThrowStatement
-            // We only care about casting in the former case.
+            // We only care about casting in the former case since the two
+            // types being coalesce-d might not be the same and might result in broken
+            // code without the cast.
+            // In the latter case something like
+            // _ = myParameter ?? throw new ArgumentNullException(nameof(myParameter));
+            // will be always valid.
             if (!syntaxFacts.IsSimpleAssignmentStatement(whenTrueStatement))
                 return expressionToCoalesce;
 

--- a/src/Analyzers/Core/CodeFixes/UseCoalesceExpression/AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
+++ b/src/Analyzers/Core/CodeFixes/UseCoalesceExpression/AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
@@ -26,11 +26,10 @@ internal abstract class AbstractUseCoalesceExpressionForIfNullStatementCheckCode
         return Task.CompletedTask;
     }
 
-    protected abstract bool ShouldAddExplicitCast(
+    protected virtual ITypeSymbol? TryGetExplicitCast(
         ISyntaxFactsService syntaxFacts, SemanticModel semanticModel,
         SyntaxNode expressionToCoalesce, SyntaxNode whenTrueStatement,
-        [NotNullWhen(true)] out ITypeSymbol? castTo,
-        CancellationToken cancellationToken);
+        CancellationToken cancellationToken) => null;
 
     protected override async Task FixAllAsync(
         Document document, ImmutableArray<Diagnostic> diagnostics,
@@ -48,8 +47,8 @@ internal abstract class AbstractUseCoalesceExpressionForIfNullStatementCheckCode
 
             var left = expressionToCoalesce.WithoutTrivia();
 
-            if (ShouldAddExplicitCast(syntaxFacts, semanticModel, expressionToCoalesce,
-                whenTrueStatement, out var castTo, cancellationToken))
+            if (TryGetExplicitCast(syntaxFacts, semanticModel, expressionToCoalesce,
+                whenTrueStatement, cancellationToken) is { } castTo)
             {
                 left = generator.CastExpression(castTo, left);
             }

--- a/src/Analyzers/Core/CodeFixes/UseCoalesceExpression/AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
+++ b/src/Analyzers/Core/CodeFixes/UseCoalesceExpression/AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
@@ -27,8 +27,8 @@ internal abstract class AbstractUseCoalesceExpressionForIfNullStatementCheckCode
     }
 
     protected virtual ITypeSymbol? TryGetExplicitCast(
-        ISyntaxFactsService syntaxFacts, SemanticModel semanticModel,
-        SyntaxNode expressionToCoalesce, SyntaxNode whenTrueStatement,
+        SemanticModel semanticModel, SyntaxNode expressionToCoalesce,
+        SyntaxNode leftAssignmentPart, SyntaxNode rightAssignmentPart,
         CancellationToken cancellationToken) => null;
 
     protected override async Task FixAllAsync(
@@ -45,23 +45,31 @@ internal abstract class AbstractUseCoalesceExpressionForIfNullStatementCheckCode
             var ifStatement = diagnostic.AdditionalLocations[1].FindNode(getInnermostNodeForTie: true, cancellationToken);
             var whenTrueStatement = diagnostic.AdditionalLocations[2].FindNode(getInnermostNodeForTie: true, cancellationToken);
 
-            var left = expressionToCoalesce.WithoutTrivia();
-
-            if (TryGetExplicitCast(syntaxFacts, semanticModel, expressionToCoalesce,
-                whenTrueStatement, cancellationToken) is { } castTo)
-            {
-                left = generator.CastExpression(castTo, left);
-            }
-
             editor.RemoveNode(ifStatement);
             editor.ReplaceNode(
                 expressionToCoalesce,
                 generator.CoalesceExpression(
-                    left,
+                    TryAddExplicitCast(expressionToCoalesce, whenTrueStatement).WithoutTrivia(),
                     GetWhenNullExpression(whenTrueStatement).WithoutTrailingTrivia()).WithTriviaFrom(expressionToCoalesce));
         }
 
         return;
+
+        SyntaxNode TryAddExplicitCast(SyntaxNode expressionToCoalesce, SyntaxNode whenTrueStatement)
+        {
+            if (!syntaxFacts.IsSimpleAssignmentStatement(whenTrueStatement))
+                return expressionToCoalesce;
+
+            syntaxFacts.GetPartsOfAssignmentStatement(whenTrueStatement, out var left, out var right);
+
+            var castTo = TryGetExplicitCast(semanticModel, expressionToCoalesce, left, right, cancellationToken);
+            if (castTo is null)
+            {
+                return expressionToCoalesce;
+            }
+
+            return generator.CastExpression(castTo, expressionToCoalesce);
+        }
 
         SyntaxNode GetWhenNullExpression(SyntaxNode whenTrueStatement)
         {

--- a/src/Analyzers/Core/CodeFixes/UseCoalesceExpression/AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
+++ b/src/Analyzers/Core/CodeFixes/UseCoalesceExpression/AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
@@ -57,6 +57,8 @@ internal abstract class AbstractUseCoalesceExpressionForIfNullStatementCheckCode
 
         SyntaxNode TryAddExplicitCast(SyntaxNode expressionToCoalesce, SyntaxNode whenTrueStatement)
         {
+            // This can be either SimpleAssignmentStatement or ThrowStatement
+            // We only care about casting in the former case.
             if (!syntaxFacts.IsSimpleAssignmentStatement(whenTrueStatement))
                 return expressionToCoalesce;
 

--- a/src/Analyzers/Core/CodeFixes/UseCoalesceExpression/AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
+++ b/src/Analyzers/Core/CodeFixes/UseCoalesceExpression/AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.cs
@@ -2,27 +2,20 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System;
 using System.Collections.Immutable;
-using System.Composition;
+using System.Diagnostics.CodeAnalysis;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.CodeAnalysis.CodeActions;
 using Microsoft.CodeAnalysis.CodeFixes;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Editing;
-using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.LanguageService;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.UseCoalesceExpression;
 
-[ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic, Name = PredefinedCodeFixProviderNames.UseCoalesceExpressionForIfNullStatementCheck), Shared]
-[ExtensionOrder(Before = PredefinedCodeFixProviderNames.AddBraces)]
-[method: ImportingConstructor]
-[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-internal class UseCoalesceExpressionForIfNullStatementCheckCodeFixProvider() : SyntaxEditorBasedCodeFixProvider
+internal abstract class AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider() : SyntaxEditorBasedCodeFixProvider
 {
     public override ImmutableArray<string> FixableDiagnosticIds
         => [IDEDiagnosticIds.UseCoalesceExpressionForIfNullCheckDiagnosticId];
@@ -33,10 +26,17 @@ internal class UseCoalesceExpressionForIfNullStatementCheckCodeFixProvider() : S
         return Task.CompletedTask;
     }
 
-    protected override Task FixAllAsync(
+    protected abstract bool ShouldAddExplicitCast(
+        ISyntaxFactsService syntaxFacts, SemanticModel semanticModel,
+        SyntaxNode expressionToCoalesce, SyntaxNode whenTrueStatement,
+        [NotNullWhen(true)] out ITypeSymbol? castTo,
+        CancellationToken cancellationToken);
+
+    protected override async Task FixAllAsync(
         Document document, ImmutableArray<Diagnostic> diagnostics,
         SyntaxEditor editor, CancellationToken cancellationToken)
     {
+        var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
         var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
         var generator = editor.Generator;
 
@@ -46,15 +46,23 @@ internal class UseCoalesceExpressionForIfNullStatementCheckCodeFixProvider() : S
             var ifStatement = diagnostic.AdditionalLocations[1].FindNode(getInnermostNodeForTie: true, cancellationToken);
             var whenTrueStatement = diagnostic.AdditionalLocations[2].FindNode(getInnermostNodeForTie: true, cancellationToken);
 
+            var left = expressionToCoalesce.WithoutTrivia();
+
+            if (ShouldAddExplicitCast(syntaxFacts, semanticModel, expressionToCoalesce,
+                whenTrueStatement, out var castTo, cancellationToken))
+            {
+                left = generator.CastExpression(castTo, left);
+            }
+
             editor.RemoveNode(ifStatement);
             editor.ReplaceNode(
                 expressionToCoalesce,
                 generator.CoalesceExpression(
-                    expressionToCoalesce.WithoutTrivia(),
+                    left,
                     GetWhenNullExpression(whenTrueStatement).WithoutTrailingTrivia()).WithTriviaFrom(expressionToCoalesce));
         }
 
-        return Task.CompletedTask;
+        return;
 
         SyntaxNode GetWhenNullExpression(SyntaxNode whenTrueStatement)
         {

--- a/src/Analyzers/VisualBasic/CodeFixes/UseCoalesceExpression/VisualBasicUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.vb
+++ b/src/Analyzers/VisualBasic/CodeFixes/UseCoalesceExpression/VisualBasicUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.vb
@@ -2,7 +2,6 @@
 ' The .NET Foundation licenses this file to you under the MIT license.
 ' See the LICENSE file in the project root for more information.
 
-
 Imports System.Composition
 Imports System.Diagnostics.CodeAnalysis
 Imports System.Threading

--- a/src/Analyzers/VisualBasic/CodeFixes/UseCoalesceExpression/VisualBasicUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.vb
+++ b/src/Analyzers/VisualBasic/CodeFixes/UseCoalesceExpression/VisualBasicUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.vb
@@ -20,9 +20,5 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UseCoalesceExpression
         <Obsolete(MefConstruction.ImportingConstructorMessage, True)>
         Public Sub New()
         End Sub
-
-        Protected Overrides Function ShouldAddExplicitCast(syntaxFacts As ISyntaxFactsService, semanticModel As SemanticModel, expressionToCoalesce As SyntaxNode, whenTrueStatement As SyntaxNode, <NotNullWhen(True)> ByRef castTo As ITypeSymbol, cancellationToken As CancellationToken) As Boolean
-            Return False
-        End Function
     End Class
 End Namespace

--- a/src/Analyzers/VisualBasic/CodeFixes/UseCoalesceExpression/VisualBasicUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.vb
+++ b/src/Analyzers/VisualBasic/CodeFixes/UseCoalesceExpression/VisualBasicUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.vb
@@ -1,0 +1,29 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
+' The .NET Foundation licenses this file to you under the MIT license.
+' See the LICENSE file in the project root for more information.
+
+
+Imports System.Composition
+Imports System.Diagnostics.CodeAnalysis
+Imports System.Threading
+Imports Microsoft.CodeAnalysis.CodeFixes
+Imports Microsoft.CodeAnalysis.Host.Mef
+Imports Microsoft.CodeAnalysis.LanguageService
+Imports Microsoft.CodeAnalysis.UseCoalesceExpression
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.UseCoalesceExpression
+    <ExportCodeFixProvider(LanguageNames.VisualBasic, Name:=PredefinedCodeFixProviderNames.UseCoalesceExpressionForIfNullStatementCheck), [Shared]>
+    <ExtensionOrder(Before:=PredefinedCodeFixProviderNames.AddBraces)>
+    Friend Class VisualBasicUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider
+        Inherits AbstractUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider
+
+        <ImportingConstructor>
+        <Obsolete(MefConstruction.ImportingConstructorMessage, True)>
+        Public Sub New()
+        End Sub
+
+        Protected Overrides Function ShouldAddExplicitCast(syntaxFacts As ISyntaxFactsService, semanticModel As SemanticModel, expressionToCoalesce As SyntaxNode, whenTrueStatement As SyntaxNode, <NotNullWhen(True)> ByRef castTo As ITypeSymbol, cancellationToken As CancellationToken) As Boolean
+            Return False
+        End Function
+    End Class
+End Namespace

--- a/src/Analyzers/VisualBasic/CodeFixes/UseCoalesceExpression/VisualBasicUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.vb
+++ b/src/Analyzers/VisualBasic/CodeFixes/UseCoalesceExpression/VisualBasicUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.vb
@@ -3,11 +3,8 @@
 ' See the LICENSE file in the project root for more information.
 
 Imports System.Composition
-Imports System.Diagnostics.CodeAnalysis
-Imports System.Threading
 Imports Microsoft.CodeAnalysis.CodeFixes
 Imports Microsoft.CodeAnalysis.Host.Mef
-Imports Microsoft.CodeAnalysis.LanguageService
 Imports Microsoft.CodeAnalysis.UseCoalesceExpression
 
 Namespace Microsoft.CodeAnalysis.VisualBasic.UseCoalesceExpression

--- a/src/Analyzers/VisualBasic/CodeFixes/VisualBasicCodeFixes.projitems
+++ b/src/Analyzers/VisualBasic/CodeFixes/VisualBasicCodeFixes.projitems
@@ -50,6 +50,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)SimplifyLinqExpression\VisualBasicSimplifyLinqExpressionCodeFixProvider.vb" />
     <Compile Include="$(MSBuildThisFileDirectory)SimplifyObjectCreation\VisualBasicSimplifyObjectCreationCodeFixProvider.vb" />
     <Compile Include="$(MSBuildThisFileDirectory)UnsealClass\VisualBasicUnsealClassCodeFixProvider.vb" />
+    <Compile Include="$(MSBuildThisFileDirectory)UseCoalesceExpression\VisualBasicUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider.vb" />
     <Compile Include="$(MSBuildThisFileDirectory)UseCollectionInitializer\VisualBasicUseCollectionInitializerCodeFixProvider.vb" />
     <Compile Include="$(MSBuildThisFileDirectory)UseCompoundAssignment\VisualBasicUseCompoundAssignmentCodeFixProvider.vb" />
     <Compile Include="$(MSBuildThisFileDirectory)UseConditionalExpression\MultiLineConditionalExpressionFormattingRule.vb" />

--- a/src/Analyzers/VisualBasic/Tests/UseCoalesceExpression/UseCoalesceExpressionForIfNullStatementCheckTests.vb
+++ b/src/Analyzers/VisualBasic/Tests/UseCoalesceExpression/UseCoalesceExpressionForIfNullStatementCheckTests.vb
@@ -10,7 +10,7 @@ Imports Microsoft.CodeAnalysis.VisualBasic.UseCoalesceExpression
 
 Imports VerifyVB = Microsoft.CodeAnalysis.Editor.UnitTests.CodeActions.VisualBasicCodeFixVerifier(Of
     Microsoft.CodeAnalysis.VisualBasic.UseCoalesceExpression.VisualBasicUseCoalesceExpressionForIfNullStatementCheckDiagnosticAnalyzer,
-    Microsoft.CodeAnalysis.UseCoalesceExpression.UseCoalesceExpressionForIfNullStatementCheckCodeFixProvider)
+    Microsoft.CodeAnalysis.VisualBasic.UseCoalesceExpression.VisualBasicUseCoalesceExpressionForIfNullStatementCheckCodeFixProvider)
 
 Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.UseCoalesceExpression
     <Trait(Traits.Feature, Traits.Features.CodeActionsUseCoalesceExpression)>
@@ -252,6 +252,88 @@ end class
             Await New VerifyVB.Test With {
                 .TestCode = text,
                 .FixedCode = text
+            }.RunAsync()
+        End Function
+
+        <Fact, WorkItem("https://github.com/dotnet/roslyn/issues/74460")>
+        Public Async Function TestLocalDeclaration_NoCastBaseAssignment() As Task
+            Await New VerifyVB.Test With {
+                .TestCode = "
+interface I
+end interface
+
+class C
+    implements I
+
+    sub Main(o as object)
+        dim item as I = TryCast(o, C)
+        [|if|] item is nothing then
+            item = TryCast(o, D)
+        end if
+    end sub
+end class
+
+class D
+    implements I
+end class
+                ",
+                .FixedCode = "
+interface I
+end interface
+
+class C
+    implements I
+
+    sub Main(o as object)
+        dim item as I = If(TryCast(o, C), TryCast(o, D))
+    end sub
+end class
+
+class D
+    implements I
+end class
+                "
+            }.RunAsync()
+        End Function
+
+        <Fact, WorkItem("https://github.com/dotnet/roslyn/issues/74460")>
+        Public Async Function TestLocalDeclaration_NoCastBaseAssignment1() As Task
+            Await New VerifyVB.Test With {
+                .TestCode = "
+interface I
+end interface
+
+class C
+    implements I
+
+    sub Main(c as C, d as D)
+        dim item as I = c
+        [|if|] item is nothing then
+            item = d
+        end if
+    end sub
+end class
+
+class D
+    implements I
+end class
+                ",
+                .FixedCode = "
+interface I
+end interface
+
+class C
+    implements I
+
+    sub Main(c as C, d as D)
+        dim item as I = If(c, d)
+    end sub
+end class
+
+class D
+    implements I
+end class
+                "
             }.RunAsync()
         End Function
     End Class

